### ISSUE TITLE
Extracting word embeddings from FlauBERT

### ISSUE THREAD

**Issue by mcriggs** [getalp/Flaubert#37]
Hello,

I would like to use FlauBERT to investigate semantic ambiguity in eighteenth-century French. To do so, I would like to extract the word embeddings of a specific keyword from a large group of different sentences in which the keyword was used and visualize those word embeddings. I have been able to extract and visualize word embeddings using BERT (after having looked at various online tutorials), but I cannot figure out how to revise the code I've been using so that it works with FlauBERT. I've inserted the code I have used for extracting and visualizing word embeddings using BERT. Could you possibly give me some suggestions as to how I might modify this code so that it works with FlauBERT?

I also wonder whether I should further train FlauBERT on a corpus of eighteenth-century French texts?

Thank you in advance for your time.

```
!pip install torch
!pip install transformers

from transformers import BertTokenizer, BertModel
import pandas as pd
import numpy as np
import nltk
import torch
from sklearn.decomposition import PCA
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import LeaveOneOut
from sklearn import model_selection
import matplotlib.pyplot as plt
%matplotlib inline

text_data = pd.read_csv("https://raw.githubusercontent.com/name/file.csv")
texts = text_data['Sentence'].to_list()

model = BertModel.from_pretrained('bert-base-uncased',
                                  output_hidden_states = True,
                                  )
tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')

def bert_text_preparation(text, tokenizer):
    
    marked_text = "[CLS] " + text + " [SEP]"
    tokenized_text = tokenizer.tokenize(marked_text)
    indexed_tokens = tokenizer.convert_tokens_to_ids(tokenized_text)
    segments_ids = [1]*len(indexed_tokens)

    tokens_tensor = torch.tensor([indexed_tokens])
    segments_tensors = torch.tensor([segments_ids])

    return tokenized_text, tokens_tensor, segments_tensors

def get_bert_embeddings(tokens_tensor, segments_tensors, model):

    with torch.no_grad():
        outputs = model(tokens_tensor, segments_tensors)
        hidden_states = outputs[2][1:]

    token_embeddings = hidden_states[-1]
    token_embeddings = torch.squeeze(token_embeddings, dim=0)
    list_token_embeddings = [token_embed.tolist() for token_embed in token_embeddings]

    return list_token_embeddings

target_word_embeddings = []

for text in texts:
    tokenized_text, tokens_tensor, segments_tensors = bert_text_preparation(text, tokenizer)
    list_token_embeddings = get_bert_embeddings(tokens_tensor, segments_tensors, model)
    
    word_index = tokenized_text.index('keyword')
    word_embedding = list_token_embeddings[word_index]

    target_word_embeddings.append(word_embedding)

keyword_pca = PCA(n_components=2).fit_transform(target_word_embeddings)
keyword_pca.shape

scatter_x = keyword_pca[:,0]
scatter_y = keyword_pca[:,1]
fig, ax = plt.subplots(figsize=(10, 7))
for g in np.unique(text_data.Place_holder):
    ix = np.where(text_data.Place_holder == g)
    ax.scatter(scatter_x[ix], scatter_y[ix],s=100)
plt.show()
```

**Comment by formiel** [comment]
Hello @mcriggs,

Thank you for your interest in FlauBERT and sorry for my late response! 

To generate the embeddings for FlauBERT, you can use either the HuggingFace's `transformers` library or `XLM` one. We provided examples for doing so in [here](https://github.com/getalp/Flaubert#2-using-flaubert). 

To continue training FlauBERT on your own corpus, you can run the [following](https://github.com/getalp/Flaubert#32-training) training command with either of the 2 below additional options:
- (1) `--reload_model`: this option will initialize the model with the pre-trained weights.
- (2) `--reload_checkpoint`: this option will resume training from the last checkpoint (so the learning rate will be computed based on the one from the last epoch).
We provided both model weights and checkpoints [here](https://github.com/getalp/Flaubert#1-flaubert-models).

Please feel free to let me know if you have any issue in using FlauBERT.

**Comment by mcriggs** [comment]
Thank you for your helpful response! 
I have a follow-up question. If I were looking to extract the word embedding for a particular word (say, for example, the word "chat" in the example you provide), would it make sense to do that as follows:

```
modelname = 'flaubert/flaubert_base_cased' 
flaubert, log = FlaubertModel.from_pretrained(modelname, output_loading_info=True)
flaubert_tokenizer = FlaubertTokenizer.from_pretrained(modelname, do_lowercase=False)

sentence = "Le chat mange une pomme."
token_ids = torch.tensor([flaubert_tokenizer.encode(sentence)])
last_layer = flaubert(token_ids)[0]

#embedding for "chat".
sentence_to_tokenize = "x " + sentence + " x"
tokenized_sentence = flaubert_tokenizer.tokenize(sentence_to_tokenize)
tokenized_sentence.index('chat</w>')
chat_word_index = tokenized_sentence.index('chat</w>')
chat_word_embedding = last_layer[:, word_index, :]
```
